synapse/handlers/sync.py

```diff
@@ -1378,13 +1378,27 @@ async def generate_sync_result(
                 else:
                     mutable_joined_room_ids.discard(room_id)
 
+        # Tweak the set of rooms to return to the client for eager (non-lazy) syncs.
         mutable_rooms_to_exclude = set(self.rooms_to_exclude_globally)
+        retransmit_room_ids = set()
         if not sync_config.filter_collection.lazy_load_members():
+            # Non-lazy syncs should never include partially stated rooms.
             # Exclude all partially stated rooms from this sync.
             for room_id in mutable_joined_room_ids:
                 if await self.store.is_partial_state_room(room_id):
                     mutable_rooms_to_exclude.add(room_id)
 
+            # Incremental syncs should include rooms that became fully-stated during
+            # the sync period.
+            if since_token:
+                retransmit_room_ids.update(
```

Comment:
Erik: `retransmit_room_ids` should not contain anything in `mutable_rooms_to_exclude`. (Join/leave/join/leave)

Reply:
c258e7298d0f58906328a66e41f53194979f5db5